atsign-foundation/at_c
# test: CRUD functional_tests, atchops fixes, & misc changes

Closes #102 
Closes #177 

## What I did
NEW Functional Tests:
- New `.github/workflows/tests.yaml` -> Now runs unit tests AND functional tests. (Deleted unit_tests.yaml)
- GHA Secrets has 2 atKeys: ATKEYS_12ALPACA and ATKEYS_12SNOWBOATING
- New `tests/functional_tests/` folder
- New `tools/run_ctest.sh` to run the CTest.
- New config.h file which has configurable root host/port and CI atSigns.
- New config.c.  Tests expect atKeys to be in ~/.atsign/keys/
- New helpers.c and helpers.h ; contains helper functions for functional tests
- New: 3 functional test suites: `test_atclient_sharedkey.c`, `test_atclient_publickey.c` and `test_atclient_sharedkey.c` all in `tests/functional_tests/tests/`

atchops changes:
- Fixed flaking AES CTR unit test (test_aes_generate.c)
- Made RSA functions to strictly only work with SHA 256 (only works for supported SHA256 which is currently the only tested MD type in the SDK, am unsure about other MD alogirthms and RSA1024, which is why I've made those functions work only for SHA256 and RSA2048).

Misc changes:
- Renamed `atclient_atlogger_log` to `atlogger_log`
- Renamed valuelen to valuesize and valueolen to valuelen in some functions
- Removed root_conn from the function headers that don't use it
- Optimized pkam_authenticate by 2KB
- Fixes to atclient_put not working

## How I did it
- With a lot of trial and error

## How to verify it
- See Functional Tests in GHA
- Did manual tests on desktop examples as of commit: https://github.com/atsign-foundation/at_c/pull/210/commits/a16942e95002919e37332d5c8347bfedef1dc865
